File: Alert.cpp
#include "Alert.hpp"

// Конструктор
Alert::Alert(const std::string& id,
    const std::string& sensorId,
    const std::string& message,
    Severity severity)
    : m_id(id),
    m_sensorId(sensorId),
    m_message(message),
    m_severity(severity)
{
    std::cout << "Alert created: " << m_message << std::endl;
}

// Деструктор
Alert::~Alert()
{
    std::cout << "Alert destroyed: " << m_id << std::endl;
}

// Уведомление текстом
void Alert::SendTextNotification() const
{
    std::cout << "[TEXT ALERT] "
        << m_message
        << " | Level: "
        << SeverityToString(m_severity)
        << std::endl;
}

// Уведомление звуком
void Alert::SendSoundNotification() const
{
    std::cout << "[SOUND ALERT] *beep* "
        << m_message << std::endl;
}

// Преобразование enum → строка
std::string Alert::SeverityToString(Severity severity)
{
    switch (severity)
    {
    case Severity::INFO:     return "INFO";
    case Severity::WARNING:  return "WARNING";
    case Severity::CRITICAL: return "CRITICAL";
    }
    return "UNKNOWN";
}

File: Lab_04/Lab04/Lab04Monitoring/Alert.cs
// Класс Alert описывает одно оповещение для пользователя.
// Здесь мы используем enum Severity и свойства вместо публичных полей.
using System;

namespace Lab04Monitoring
{
    public class Alert
    {
        // Идентификатор оповещения.
        public string Id { get; }

        // Идентификатор датчика-источника.
        public string SensorId { get; }

        // Текст сообщения.
        public string Message { get; }

        // Уровень серьёзности.
        public Severity Severity { get; }

        // Конструктор обычный.
        public Alert(string id, string sensorId, string message, Severity severity)
        {
            Id = id;
            SensorId = sensorId;
            Message = message;
            Severity = severity;
        }

        // Текстовое оповещение в консоль.
        public void SendTextNotification()
        {
            Console.WriteLine($"[ALERT {Id}] ({Severity}) from sensor {SensorId}: {Message}");
        }

        // "Звуковое" оповещение для демонстрации.
        public void SendSoundNotification()
        {
            Console.WriteLine("[SOUND ALERT] *beep* " + Message);
        }

        // Переопределяем ToString, чтобы удобно использовать Alert в логах.
        public override string ToString()
        {
            return $"Alert(Id={Id}, Sensor={SensorId}, Severity={Severity}, Message=\"{Message}\")";
        }
    }
}
